### PR TITLE
Add extension command module

### PR DESCRIPTION
**Pull request by calebkiage**
## Overview

Adds a command to manage 3rd party extensions in the cli

### Demo

```
(msgraph-cli) > ./mg extension list
[]

(msgraph-cli) >./mg extension show
ERROR: the following arguments are required: --name/-n

TRY THIS:
az extension show --name MyExtension
Show an extension. (autogenerated)

az extension list
List the installed extensions. (autogenerated)

https://docs.microsoft.com/en-US/cli/azure/extension#az_extension_show
Read more about the command in reference docs

(msgraph-cli) >./mg extension list-available
[
  {
    "experimental": false,
    "installed": false,
    "name": "mail_beta",
    "preview": true,
    "summary": "Microsoft Azure Command-Line Tools Mail Extension",
    "version": "0.1.0"
  }
]

```

## Testing Instructions

* Run the command `./mg extension --help` to see the available commands
* Run any of the commands to test.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoftgraph/msgraph-cli/pull/153)

### CODE REVIEW

File: msgraph/cli/command_modules/extension/_help.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+helps['extension add'] = """
+type: command
+short-summary: Add an extension.
+long-summary: To learn about installing extensions automatically, visit https://aka.ms/AzExtDynamicInstall.
+parameters:
+  - name: --system
+    type: string
+    short-summary: Use a system directory for the extension.
+    long-summary: Default path is azure-cli-extensions folder under the CLI running python environment lib path, configurable by environment variable AZURE_EXTENSION_SYS_DIR. On Windows, you may need to open your shell as Administrator to run with the right permission.
```
Comment (samwelkanda) [inline msgraph/cli/command_modules/extension/_help.py:23]: We need to replace `azure` references with `msgraph` in the help strings

File: msgraph/cli/command_modules/extension/_help.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+examples:
+  - name: Add extension by name
+    text: az extension add --name anextension
```
Comment (samwelkanda) [inline msgraph/cli/command_modules/extension/_help.py]: Also replace `az` with `mg` in the samples.

**Review by samwelkanda — CHANGES_REQUESTED** [review]

File: msgraph/cli/command_modules/extension/_help.py (hunk 1 above)
Comment (calebkiage) [reply msgraph/cli/command_modules/extension/_help.py:23]: I haven't yet updated the extensions folder.

Comment (samwelkanda) [reply msgraph/cli/command_modules/extension/_help.py:23]: Cool let's leave it as is until we update that.

**Review by samwelkanda — APPROVED** [review]

**Review by MIchaelMainer — CHANGES_REQUESTED** [review]
I think we should clarify that these are CLI extensions. For example, in Microsoft Graph, there are schema extensions and open extensions. This could cause confusion by generically calling this extensions.

I propose something like:

```
./mg cli-extension list-available
./mg cli-extensions list-available
./mg cliextension list-available
```

I think the first option is best.

**Comment by jobala** [comment]
@MIchaelMainer I prefer `./mg extension` because it provides an interface similar to Azure CLI's and most of our users will be familiar with how Azure CLI works. I would tradeoff the confusion with consistency with Azure CLI.

**Comment by MIchaelMainer** [comment]
In terms of consistency, with the same concept in Azure, yes, keeping the name would provide consistency in the CLI experience. If we changed the name, it would still be functionally consistent, but without the same command group name as experienced in the Azure CLI. We also have to consider the term **extension** across all of Microsoft Graph:

https://docs.microsoft.com/en-us/search/?scope=graph&view=graph-rest-1.0&terms=extension

There are a lot of hits on extension.

We need to think about a future where we generate CLI code snippets and inject into the documentation. We also need to consider the impact on search indices when customers search on the term extension. Extension is already an overloaded term in Microsoft Graph (and software development in general). By not disambiguating it here, we increase the potential for confusion across the CLI, documentation, and pretty much anywhere else that Microsoft Graph based content is indexed or used.

Case in point: if the folks who created the schema extension feature called it extension, we would have had to disambiguate the term at this point. But they did the right thing and disambiguated the term so that we can have this discussion 😄.   There is little ambiguity by calling it cli-extension.

**Review by MIchaelMainer — APPROVED** [review]

File: Pipfile [hunk 3]
```diff
@@ -6,13 +6,17 @@ verify_ssl = true
...
 bump2version = "*"
+azure-storage-blob = "==2.1.0"
 
 [packages]
-azure-cli-core = {path = "./../azure-cli/src/azure-cli-core"}
\ No newline at end of file
+azure-cli-core = { git = "https://github.com/microsoftgraph/azure-cli.git", ref = "28d820dc748d37df0ff70fed45c69d7168a656d3", subdirectory = "src/azure-cli-core" }
+
+[pipenv]
+allow_prereleases = true
```
Comment (MIchaelMainer) [inline Pipfile]: I'm curious why we want to support pre-releases.

File: build_scripts/automation/build_package.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    def __init__(self, name, path) -> None:
+        self.name = name
+        self.path = path
+
+
+def create_package(info, dest_folder=DEFAULT_DEST_FOLDER):
+    absdirpath = (info.path)
+    dest = os.path.abspath(dest_folder)
+    check_call([sys.executable, 'setup.py', 'bdist_wheel', '-d', dest], cwd=absdirpath)
+    # check_call(['python', 'setup.py', "sdist", "--format", "zip", '-d', dest], cwd=absdirpath)
```
Comment (MIchaelMainer) [inline build_scripts/automation/build_package.py]: nit: cleanup commented code.

File: build_scripts/automation/upload_extensions.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    updated_indexes = []
+    print('Uploading {} extensions...\n'.format(len(built_extensions.keys())))
+    # start with an empty index.json to sync all extensions
+    initial_index = {"extensions": {}, "formatVersion": "1"}
+    open(target_index_path, 'w').write(json.dumps(initial_index, indent=4, sort_keys=True))
+    for ext_name in current_extensions:
+        ext_file = built_extensions[ext_name]
+        print('Uploading {}'.format(ext_file))
+        ext_info = {}
+        _upload_wheel(ext_info=ext_info, updated_indexes=updated_indexes, ext_file=ext_file, ext_name=ext_name, client=client)
```
Comment (MIchaelMainer) [inline build_scripts/automation/upload_extensions.py:85]: Are there any cases where _upload_wheel could fail in middle of this loop resulting in updated wheels without an update to the index?

File: Pipfile (hunk 3 above)
Comment (calebkiage) [reply Pipfile:22]: I had to enable this to successfully run `pipenv install`

File: build_scripts/automation/upload_extensions.py (hunk 5 above)
Comment (calebkiage) [reply build_scripts/automation/upload_extensions.py:85]: That's a good question. Currently, this script only handles the first time run at the moment and running it a second time will overwrite the uploaded extensions. I want to update it to enable updating existing indexes on the next phase of the task. That update should handle for failures too.

**Review by jobala — APPROVED** [review]

**Review by samwelkanda — APPROVED** [review]

**Pull request merged**